GVFS/GVFS.Common/Maintenance/LooseObjectsStep.cs

```diff
@@ -0,0 +1,120 @@
+            {
+                // Check if the directory is 2 letter HEX
+                if (System.Text.RegularExpressions.Regex.IsMatch(directory, @"[0-9a-fA-F]{2}"))
+                {                    
+                    count += this.Context.FileSystem.GetFiles(Path.Combine(this.Context.Enlistment.GitObjectsRoot, directory), "*").Count();
+                }
+            }
+
+            return count;
+        }
+
+        protected override void PerformMaintenance()
+        {
+            using (ITracer activity = this.Context.Tracer.StartActivity("LooseObjectsStep", EventLevel.Informational, Keywords.Telemetry, metadata: null))
+            {
+                try
+                {
+                    if (!this.ignoreTimeRestriction && !this.EnoughTimeBetweenRuns())
+                    {
+                        activity.RelatedWarning("Skipping step due to not enough time between runs");
+                        return;
+                    }
+
+                    // Verify no git processes are running
+                    IEnumerable<int> processIds = this.RunningGitProcessIds();
+                    if (processIds.Count() > 0)
+                    {
+                        activity.RelatedWarning("Skipping LooseObjectsStep due to git pids {0}", string.Join(",", processIds));
+                        return;
+                    }
+
+                    // Perform prune-packed
+                    int beforeCount = this.CountLooseObjects();
+                    GitProcess.Result result = this.RunGitCommand((process) => process.PrunePacked(this.Context.Enlistment.GitObjectsRoot));
+                    int afterCount = this.CountLooseObjects();
+
+                    // Record Telemetry
+                    EventMetadata metadata = new EventMetadata();
+                    metadata.Add("StartingCount", beforeCount);
+                    metadata.Add("EndingCount", afterCount);
+                    metadata.Add("RemovedCount", afterCount - beforeCount);
+                    activity.RelatedEvent(EventLevel.Informational, "LooseObjects", metadata, Keywords.Telemetry);
+
+                    // Update the history file
+                    this.RecordLooseObjectHistory();
+                }
+                catch (Exception e)
+                {
+                    activity.RelatedError("Failed to run LooseObjectsStep: {0}", e.ToString());
+                }
+            }
+        }
+
+        // Job should be run 7 days apart
+        private bool EnoughTimeBetweenRuns()
+        {
+            if (!this.Context.FileSystem.FileExists(this.looseObjectsHistoryPath))
+            {
+                return true;
+            }
+
+            string lastRunTime = this.Context.FileSystem.ReadAllText(this.looseObjectsHistoryPath);
+            DateTime dateTime = Convert.ToDateTime(lastRunTime);
+            double daysDiff = (DateTime.Now - dateTime).TotalDays;
+            if (daysDiff >= 7)

```

Comment:
This seems like it would be better to use a `readonly TimeSpan` up with the constants, and then compare the values:

```
private readonly TimeSpan timeBetweenRuns = TimeSpan.FromDays(7);
```
...
```
            string lastRunTime = this.Context.FileSystem.ReadAllText(this.looseObjectsHistoryPath);
            DateTime lastRunDate = Convert.ToDateTime(lastRunTime);
            TimeSpan timeSince = DateTime.Now - lastRunDate;
            if (timeSince.CompareTo(this.TimeBetweenRuns) > 0)
            {
                return true;
            }
```

Also, this logic is not specific to this step, but will be something we want to replicate in other steps. Specifically:

1. We should remove the dependence in the `PrefetchStep` to use the filenames for timestamps and instead use this file mechanism. (This may require changing the self-healing logic present currently, so is a bigger job than it sounds.) The prefetch step will want to actually acquire this timestamp as a Unix Epoch value.
2. The repack step will need _exactly_ this logic, but with a different filename.

Related to all of this: we have `EpochConverter` to assist with using Unix-style epoch values instead of `DateTime` conversions (which may have localization issues?).

And finally, my super nit-pick: "history" doesn't seem like the right word, as we don't keep a log of multiple runs. Perhaps `.time` as the extension?

Reply:
@derrickstolee 

I've updated the timestamp reading and writing to a file to be in MaintenanceStep.  This gives steps the opportunity to use it.  I wouldn't however move PrefetchStep over to the new infrastructure in this design since it's a bit involved and this PR is already quite large.  I'd rather create an issue and seperate PR for it.  Sound fair?

I'm not sure why we'd need to use EpochConverter.  Since it's reading/writing from the same machine shouldn't any localization be accounted for?

Reply:
Times are hard. Localization is hard. I never claim to know handling them will work.

For instance: what happens if someone updates their timezone? How will that interact with your DateTime serializations? No, it's not the end of the world if that happens, but an epoch is simple and related to our existing timestamp patterns.

Reply:
@derrickstolee I think we can fix the timezone issue and keep the data readable by using Utc and a date format.  Let know what you think about the latest changes.